bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/RecoveryManager.java

```diff
@@ -0,0 +1,315 @@
+                    // if the transaction was in any of the terminating states, it means that the 2pc has initiated
+                    // and, it has impacted the resources. Therefore, we need to recover the transaction accordingly.
+                    Xid xid = XIDGenerator.createXID(logRecord.getCombinedId());
+                    boolean singleTrxRecoverSuccess =
+                            recoverFailedTrxInAllResources(xid, logRecord.getTransactionState());
+                    if (singleTrxRecoverSuccess) {
+                        // put a terminated log record to indicate that the transaction was recovered successfully
+                        TransactionLogRecord terminatedRecord = new TransactionLogRecord(
+                                logRecord.getTransactionId(), logRecord.getTransactionBlockId(),
+                                RecoveryState.TERMINATED);
+                        TransactionResourceManager.getInstance().getLogManager().put(terminatedRecord);
+                        iterator.remove();
+                    }
+                    recoverSuccess = recoverSuccess && singleTrxRecoverSuccess;
+                }
+                case MIXED, HAZARD -> {
+                    // if the transaction was in any of the mixed or hazard states, it means that the transaction is
+                    // not in a state we can handle and should be recovered manually, so we inform the user.
+                    String combinedId = logRecord.getCombinedId();
+                    switch (logRecord.getTransactionState()) {
+                        case MIXED -> diagnosticLog.warn(ErrorCodes.TRANSACTION_IN_MIXED_STATE, null, combinedId);
+                        case HAZARD -> diagnosticLog.warn(ErrorCodes.TRANSACTION_IN_HAZARD_STATE, null, combinedId);
+                    }
+                    iterator.remove(); // consider it handled, as we have warned the user
+                }
+            }
+        }
+
+        if (!recoverSuccess) {
+            diagnosticLog.error(ErrorCodes.TRANSACTION_STARTUP_RECOVERY_FAILED, null);
+        }
+
+        // notify the user of all startup recovery warns and errors
+        if (!diagnosticLog.getDiagnosticList().isEmpty()) {
+            RuntimeUtils.handleDiagnosticErrors(diagnosticLog);
+        }
+        return recoverSuccess;
+    }
+
+    /**
+     * Recovers a failed transactions in all XAResources.
+     *
+     * @param xid   the xid of the transaction
+     * @param state the state of the transaction
+     * @return true if all transactions are recovered successfully from all resources, false otherwise
+     */
+    private boolean recoverFailedTrxInAllResources(Xid xid, RecoveryState state) {
+        boolean allResourcesRecovered = true;
+        for (XAResource xaResource : xaResources) {
+            allResourcesRecovered = allResourcesRecovered && recoverFailedTrxInXAResource(xaResource, xid, state);
+        }
+        return allResourcesRecovered;
```

Comment:
@heshanpadmasiri Here I want to ensure the recovery attempt is made for all resources, regardless of whether one fails. 
I want to ensure that `recoverFailedTrxInXAResource()` is evaluated in all iterations, while `allResourcesRecovered` remains false only if any iteration returns false.

Shouldn't the below change suffice that need?
```java
allResourcesRecovered = recoverFailedTrxInXAResource(xaResource, xid, state) && allResourcesRecovered;
```

Reply:
I see in that case what you have purposed is sufficient (since java [guarantees](https://docs.oracle.com/javase/specs/jls/se22/html/jls-15.html#jls-15.23) order of evaluation to be left to right). However personally I would breakup the call to `recoverFailedTrxInXAResource` and updating the `allResourcesRecovered` into two to make the fact `recoverFailedTrxInXAResource` needs to be run on all and has no effect on the flag more explicit. 

```java
boolean recoveredResource = recoverFailedTrxInXAResource(xaResource, xid, state);
allResourcesRecovered = recoveredResource && allResourcesRecovered;
```

Reply:
Yes, I agree.

Reply:
Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/42080/commits/927c6b5403944fd87d7e7f15edcb0888cdf5a606